Bundler tests failing by the `git branch -l <branch name>` creating a new branch on old Git

<!--

Thank you for contributing to the [rubygems](https://github.com/rubygems/rubygems) repository, and specifically to the [Bundler](https://bundler.io/) gem.

Before opening your issue, make sure you have checked [our filing issues guide](https://github.com/rubygems/rubygems/blob/master/bundler/doc/contributing/ISSUES.md).

Please fill in the following sections so we can process your issue as fast as possible:

-->

### Describe the problem as clearly as you can

<!-- Replace this with an explanation of the problem you are having. Be as much clear and precise as you can. -->

When running `make test-bundler` in ruby/ruby repo (Ruby 2.7.4 + manually ported Bundler 2.2.4) with old Git version `1.8.3.1` on RHEL 7, I see some tests failing on ruby/ruby `make test-bundler`.

```
$ make test-bundler 2>&1 | tee make_test-bundler.log
...
Finished in 75 minutes 17 seconds (files took 1.16 seconds to load)
2901 examples, 14 failures, 25 pending

Failed examples:

rspec ./spec/bundler/commands/add_spec.rb:108 # bundle add with --git and --branch adds dependency with specified github source and branch
rspec ./spec/bundler/commands/info_spec.rb:122 # bundle info with a git repo in the Gemfile prints out branch names other than master
rspec ./spec/bundler/commands/show_spec.rb:106 # bundle show with a git repo in the Gemfile prints out branch names other than master
rspec ./spec/bundler/install/allow_offline_install_spec.rb:72 # bundle install with :allow_offline_install with cached data locally will install from a cached git repo
rspec ./spec/bundler/install/gemfile/git_spec.rb:219 # bundle install with git sources when specifying a revision works when the revision is a non-head ref
rspec ./spec/bundler/install/gemfile/git_spec.rb:247 # bundle install with git sources when specifying a revision works when the revision is a non-head ref and it was previously downloaded
rspec ./spec/bundler/install/gemfile/git_spec.rb:305 # bundle install with git sources when specifying a branch works
rspec ./spec/bundler/install/gemfile/git_spec.rb:319 # bundle install with git sources when specifying a branch when the branch starts with a `#` works
rspec ./spec/bundler/install/gemfile/git_spec.rb:336 # bundle install with git sources when specifying a branch when the branch includes quotes works
rspec ./spec/bundler/install/gemfile/git_spec.rb:566 # bundle install with git sources when specifying local override explodes on different branches on install
rspec ./spec/bundler/install/gemfile/git_spec.rb:1324 # bundle install with git sources with an extension does reinstall the extension when changing refs
rspec ./spec/bundler/lock/lockfile_spec.rb:484 # the lockfile format generates a lockfile with a ref for a single pinned source, git gem with a branch requirement
rspec ./spec/bundler/update/git_spec.rb:5 # bundle update git sources floats on a branch when :branch is used
rspec ./spec/bundler/update/git_spec.rb:38 # bundle update git sources floats on a branch when :branch is used and the source is specified in the update

make: *** [yes-test-bundler] Error 1
```

Only running one of the failing tests, the result is like this.

```
$ make test-bundler BUNDLER_SPECS=commands/add_spec.rb:105
./revision.h unchanged
./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems -C "." bin/gem install --no-document \
	--install-dir .bundle --conservative "rspec:~> 3.5" "rake:~> 12.0" "parallel_tests:~> 2.29"
./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems -C . -Ispec/bundler .bundle/bin/rspec \
	--require spec_helper  spec/bundler/commands/add_spec.rb:105
Run options:
  include {:locations=>{"./spec/bundler/commands/add_spec.rb"=>[105]}}
  exclude {:truffleruby=>true, :jruby=>true, :readline=>false, :permissions=>false, :no_color_tty=>false, :ruby_repo=>true, :bundler=>"!= 2", :git=>"!= 1.8.3.1", :realworld=>true, :sudo=>true}
F

Failures:

  1) bundle add with --git and --branch adds dependency with specified github source and branch
     Failure/Error:
               raise <<~ERROR
       
                 Invoking `#{cmd}` failed with output:
                 ----------------------------------------------------------------------
                 #{command_execution.stdboth}
                 ----------------------------------------------------------------------
               ERROR
     
     RuntimeError:
     
     
       Commands:
       $ /builddir/build/BUILD/ruby-2.7.4/ruby -rrubygems /builddir/build/BUILD/ruby-2.7.4/bin/gem --backtrace build foo
       Successfully built RubyGem
         Name: foo
         Version: 1.1
         File: foo-1.1.gem
       # $? => 0
     
       $ /builddir/build/BUILD/ruby-2.7.4/ruby -rrubygems /builddir/build/BUILD/ruby-2.7.4/bin/gem --backtrace build foo
       Successfully built RubyGem
         Name: foo
         Version: 2.0
         File: foo-2.0.gem
       # $? => 0
     
       $ /builddir/build/BUILD/ruby-2.7.4/ruby -rrubygems /builddir/build/BUILD/ruby-2.7.4/bin/gem --backtrace build baz
       Successfully built RubyGem
         Name: baz
         Version: 1.2.3
         File: baz-1.2.3.gem
       # $? => 0
     
       $ /builddir/build/BUILD/ruby-2.7.4/ruby -rrubygems /builddir/build/BUILD/ruby-2.7.4/bin/gem --backtrace build bar
       Successfully built RubyGem
         Name: bar
         Version: 0.12.3
         File: bar-0.12.3.gem
       # $? => 0
     
       $ /builddir/build/BUILD/ruby-2.7.4/ruby -rrubygems /builddir/build/BUILD/ruby-2.7.4/bin/gem --backtrace build cat
       Successfully built RubyGem
         Name: cat
         Version: 0.12.3.pre
         File: cat-0.12.3.pre.gem
       # $? => 0
     
       $ /builddir/build/BUILD/ruby-2.7.4/ruby -rrubygems /builddir/build/BUILD/ruby-2.7.4/bin/gem --backtrace build dog
       Successfully built RubyGem
         Name: dog
         Version: 1.1.3.pre
         File: dog-1.1.3.pre.gem
       # $? => 0
     
       $ /builddir/build/BUILD/ruby-2.7.4/ruby -rrubygems /builddir/build/BUILD/ruby-2.7.4/bin/gem --backtrace generate_index
       Generating Marshal quick index gemspecs for 49 gems
       .................................................
       Complete
       Generated Marshal quick index gemspecs: 0.008s
       Generating specs index
       Generated specs index: 0.001s
       Generating latest specs index
       Generated latest specs index: 0.000s
       Generating prerelease specs index
       Generated prerelease specs index: 0.000s
       Compressing indices
       Compressed indices: 0.000s
       # $? => 0
     
       $ git config --global init.defaultBranch master
       # $? => 0
     
       $ git init
       Initialized empty Git repository in /builddir/build/BUILD/ruby-2.7.4/tmp/1/libs/foo-2.0/.git/
       # $? => 0
     
       $ git add *
       # $? => 0
     
       $ git config user.email lol@wut.com
       # $? => 0
     
       $ git config user.name lolwut
       # $? => 0
     
       $ git config commit.gpgsign false
       # $? => 0
     
       $ git commit -m OMG_INITIAL_COMMIT
       [master (root-commit) d31bb84] OMG_INITIAL_COMMIT
        3 files changed, 21 insertions(+)
        create mode 100644 foo.gemspec
        create mode 100644 lib/foo.rb
        create mode 100644 lib/foo/source.rb
       # $? => 0
     
       $ git checkout master
       Already on 'master'
       # $? => 0
     
       $ git branch -l test
       # $? => 0
     
       Invoking `git branch test` failed with output:
       ----------------------------------------------------------------------
       fatal: A branch named 'test' already exists.
       ----------------------------------------------------------------------
     # ./spec/bundler/support/helpers.rb:206:in `sys_exec'
     # ./spec/bundler/support/helpers.rb:177:in `git'
     # ./spec/bundler/support/builders.rb:559:in `_build'
     # ./spec/bundler/support/builders.rb:329:in `block in build_with'
     # ./spec/bundler/support/builders.rb:326:in `each'
     # ./spec/bundler/support/builders.rb:326:in `build_with'
     # ./spec/bundler/support/builders.rb:305:in `update_git'
     # ./spec/bundler/commands/add_spec.rb:105:in `block (3 levels) in <top (required)>'
     # ./spec/bundler/spec_helper.rb:99:in `block (4 levels) in <top (required)>'
     # ./spec/bundler/spec_helper.rb:99:in `block (3 levels) in <top (required)>'
     # ./spec/bundler/support/helpers.rb:350:in `block in with_gem_path_as'
     # ./spec/bundler/support/helpers.rb:364:in `without_env_side_effects'
     # ./spec/bundler/support/helpers.rb:346:in `with_gem_path_as'
     # ./spec/bundler/spec_helper.rb:98:in `block (2 levels) in <top (required)>'

Finished in 2.05 seconds (files took 0.17342 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/bundler/commands/add_spec.rb:108 # bundle add with --git and --branch adds dependency with specified github source and branch
```

As you know, the testing command is equivalent with the following commands on rubygems repo.

```
$ cd bundler
$ bin/rake spec:parallel_deps
$ bin/rspec spec/commands/add_spec.rb:105
```

spec/commands/add_spec.rb

```
  describe "with --git and --branch" do
    before do
      update_git "foo", "2.0", :branch => "test" # <= the line 105 is this line!
    end
    ...
```



The failure is maybe coming from the difrerence behavior of `git branch -l <branch name>`.

On the old Git 1.8.3.1, the command creates a specifified new branch.

```
+ git branch
* master
+ git branch -l test
+ git branch
* master
  test
+ git branch test
fatal: A branch named 'test' already exists.
```

On Git 2.31.1, the command doesn't create a new branch.

```
+ git branch
* master
+ git branch -l test
+ git branch
* master
+ git branch test
```

### Post steps to reproduce the problem

<!--

Fill this with a list of steps maintainers can follow to reproduce your issue. Note that while you are seeing this issue in your computer, maintainers might not see the same thing on theirs. There is a number of things that could influence this:

* How your ruby is setup (OS package, from source, using a version manager).
* How bundler & rubygems are configured.
* The version of each involved piece of software that you are using.
* ...

The more complete the steps to simulate your particular environment are, the easier it will be for maintainers to reproduce your issue on their machines.

Ideally, we recommend you to set up the list of steps as a [Dockerfile](https://docs.docker.com/get-started/). A Dockerfile provides a neutral environment that should give the same results, no matter where it's run.

-->

I prepared the reproducer on CentOS 7: https://github.com/junaruga/report-rubygems-bundler-git-error . You can run docker or podman using the reproducer repo.

```
$ docker build --rm -t test-rubygems-bundler-git .
$ docker run --rm -t test-rubygems-bundler-git
```

I wanted to create a better reproducer with `bin/rspec spec/commands/add_spec.rb:105`. However the following `bin/rake spec:parallel_deps` failed on old Ruby 2.0.0 on CentOS 7.

```
[root@e672dc9509f3 /]# ruby -v
ruby 2.0.0p648 (2015-12-16) [x86_64-linux]

[root@e672dc9509f3 /]# git clone --depth 1 https://github.com/rubygems/rubygems.git

[root@e672dc9509f3 /]# cd rubygems/bundler

[root@5df411ad9293 bundler]# bin/rake spec:parallel_deps
bin/rake:4:in `require_relative': /rubygems/bundler/spec/support/rubygems_ext.rb:75: syntax error, unexpected ',' (SyntaxError)
    def check_source_control_changes(success_message:, error_message:)
                                                      ^
/rubygems/bundler/spec/support/rubygems_ext.rb:106: syntax error, unexpected '.'
...file)).dependencies[gem_name]&.requirement
...                               ^
/rubygems/bundler/spec/support/rubygems_ext.rb:158: syntax error, unexpected keyword_end, expecting end-of-input
  from bin/rake:4:in `<main>'
```

### Which command did you run?

<!-- Replace this with the specific command that is causing trouble. -->

On ruby/ruby repo.

```
$ make test-bundler BUNDLER_SPECS=commands/add_spec.rb:105
```

### What were you expecting to happen?

<!-- Replace this with the results you expected before running the command. -->

I expect the tests pass with the old Git version 1.8.3.1.

### What actually happened?

<!-- Replace this with the actual result you got. Paste the output of your command here. -->

The tests failed.

### If not included with the output of your command, run `bundle env` and paste the output below

<!-- Replace this with the result of `bundle env`. Don't forget to anonymize any private data! -->

In this case, just running the tests without installing `bundle` command. So, I could not run `bundle env`.

--- a/bundler/spec/support/builders.rb
+++ b/bundler/spec/support/builders.rb
@@ -554,7 +554,7 @@ module Spec
           raise "You can't specify `master` as the branch" if branch == "master"
           escaped_branch = Shellwords.shellescape(branch)
 
-          if @context.git("branch -l #{escaped_branch}", libpath).empty?
+          if @context.git("branch --list #{escaped_branch}", libpath).empty?
             @context.git("branch #{escaped_branch}", libpath)
           end
 

